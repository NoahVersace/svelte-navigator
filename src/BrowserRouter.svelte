<script>
	import { browserHistory } from "svelte-navigator-history";
	import Router from "./Router.svelte";

	const UNDEF = undefined;
	export let basepath = UNDEF;
	export let url = UNDEF;
	export let primary = UNDEF;
	export let a11y = UNDEF;
	export let history = browserHistory;
</script>

<Router {a11y} {basepath} {url} {primary} {history}>
	<slot />
</Router>
