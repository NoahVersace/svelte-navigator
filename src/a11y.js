import { get } from "svelte/store";
import { warn, ROUTER_ID, ROUTE_ID } from "./warning";

// We need to keep the focus candidate in a separate file, so svelte does
// not update, when we mutate it.
// Also, we need a single global reference, because taking focus needs to
// work globally, even if we have multiple top level routers
// eslint-disable-next-line import/no-mutable-exports
export let focusCandidate = null;

// eslint-disable-next-line import/no-mutable-exports
export let initialNavigation = true;

export function pushFocusCandidate(item) {
  if (
    !focusCandidate ||
    item.level > focusCandidate.level ||
    (item.level === focusCandidate.level &&
      item.routerId < focusCandidate.routerId)
  ) {
    focusCandidate = item;
  }
}

export function clearFocusCandidate() {
  focusCandidate = null;
  initialNavigation = false;
}

/*
 * `focus` Adapted from https://github.com/oaf-project/oaf-side-effects/blob/master/src/index.ts
 *
 * https://github.com/oaf-project/oaf-side-effects/blob/master/LICENSE
 */
export function focus(elem) {
  if (!elem) return false;
  try {
    if (!elem.hasAttribute("tabindex")) {
      elem.setAttribute("tabindex", "-1");
      // We remove tabindex after blur to avoid weird browser behavior
      // where a mouse click can activate elements with tabindex="-1".
      const blurListener = () => {
        elem.removeAttribute("tabindex");
        elem.removeEventListener("blur", blurListener);
      };
      elem.addEventListener("blur", blurListener);
    }
    elem.focus();
    return document.activeElement === elem;
  } catch (e) {
    // Apparently trying to focus a disabled element in IE can throw.
    // See https://stackoverflow.com/a/1600194/2476884
    return false;
  }
}

export function resetFocus() {
  if (document.activeElement) {
    try {
      document.activeElement.blur();
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }
}

export function isEndMarker(elem, id) {
  return Number(elem.dataset.svnavRouteEnd) === id;
}

export function isHeading(elem) {
  return /^H[1-6]$/i.test(elem.tagName);
}

function query(selector, parent = document) {
  return parent.querySelector(selector);
}

export function queryHeading(id) {
  const marker = query(`[data-svnav-route-start="${id}"]`);
  let current = marker.nextElementSibling;
  while (!isEndMarker(current, id)) {
    if (isHeading(current)) {
      return current;
    }
    const heading = query("h1,h2,h3,h4,h5,h6", current);
    if (heading) {
      return heading;
    }
    current = current.nextElementSibling;
  }
  return null;
}

export function handleFocus(route) {
  Promise.resolve(get(route.focusElement)).then(elem => {
    const focusElement = elem || queryHeading(route.id);
    if (!focusElement) {
      warn(
        ROUTER_ID,
        "Could not find a heading to focus. " +
          "You should always render a header for accessibility reasons. " +
          "If you don't want this Route or Router to manage focus, " +
          // eslint-disable-next-line quotes
          'pass "primary={false}" to it.',
        route,
        ROUTE_ID,
      );
    }
    const headingFocused = focus(focusElement);
    if (headingFocused) return;
    const documentFocused = focus(document.documentElement);
    if (documentFocused) return;
    resetFocus();
  });
}

export function announceNavigation(message, announcementElement) {
  // eslint-disable-next-line no-param-reassign
  announcementElement.textContent = message;
}

export const visuallyHiddenStyle = `
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
`;
