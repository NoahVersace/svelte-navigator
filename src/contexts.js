/*
 * Adapted from https://github.com/EmilTholin/svelte-routing
 *
 * https://github.com/EmilTholin/svelte-routing/blob/master/LICENSE
 */

// Use strings instead of objects, so different versions of
// svelte-navigator can potentially still work together
export const LOCATION = "@@svnav-ctx-key__LOCATION";
export const ROUTER = "@@svnav-ctx-key__ROUTER";
export const ROUTE = "@@svnav-ctx-key__ROUTE";
