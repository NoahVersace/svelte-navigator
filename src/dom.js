import { warn, ROUTER_ID } from "./warning";

/*
 * `focus` Adapted from https://github.com/oaf-project/oaf-side-effects/blob/master/src/index.ts
 *
 * https://github.com/oaf-project/oaf-side-effects/blob/master/LICENSE
 */
export function focus(elem) {
  if (!elem) return false;
  try {
    if (!elem.hasAttribute("tabindex")) {
      elem.setAttribute("tabindex", "-1");
      // We remove tabindex after blur to avoid weird browser behavior
      // where a mouse click can activate elements with tabindex="-1".
      const blurListener = () => {
        elem.removeAttribute("tabindex");
        elem.removeEventListener("blur", blurListener);
      };
      elem.addEventListener("blur", blurListener);
    }
    elem.focus();
    return document.activeElement === elem;
  } catch (e) {
    // Apparently trying to focus a disabled element in IE can throw.
    // See https://stackoverflow.com/a/1600194/2476884
    return false;
  }
}

export function resetFocus() {
  if (document.activeElement) {
    try {
      document.activeElement.blur();
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }
}

export function isEndMarker(elem, id) {
  return Number(elem.dataset.svnavRouteEnd) === id;
}

export function isHeading(elem) {
  return /^H[1-6]$/i.test(elem.tagName);
}

function query(selector, parent = document) {
  return parent.querySelector(selector);
}

export function queryHeading(id) {
  const marker = query(`[data-svnav-route-start="${id}"]`);
  let current = marker.nextElementSibling;
  while (!isEndMarker(current, id)) {
    if (isHeading(current)) {
      return current;
    }
    const heading = query("h1,h2,h3,h4,h5,h6", current);
    if (heading) {
      return heading;
    }
    current = current.nextElementSibling;
  }
  return null;
}

export function handleFocus(route) {
  const focusHeading = queryHeading(route.id);
  if (!focusHeading) {
    const routeAttr = route.default ? "default" : `path="${route.path}"`;
    warn(
      ROUTER_ID,
      `Could not find a heading to focus in <Route ${routeAttr} />. ` +
        "You should always render a header for accessibility reasons. " +
        "If you don't want this Route or Router to manage focus, " +
        // eslint-disable-next-line quotes
        'pass "primary={false}" to it',
    );
  }
  const headingFocused = focus(focusHeading);
  if (headingFocused) return;
  const documentFocused = focus(document.documentElement);
  if (documentFocused) return;
  resetFocus();
}
