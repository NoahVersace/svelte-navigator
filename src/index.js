/*
 * Adapted from https://github.com/EmilTholin/svelte-routing
 *
 * https://github.com/EmilTholin/svelte-routing/blob/master/LICENSE
 */

import shimEnv from "./envShim";

shimEnv();

export {
	createBrowserHistory,
	createHashHistory,
	createMemoryHistory,
	browserHistory,
	hashHistory,
	memoryHistory,
} from "svelte-navigator-history";
export { default as Router } from "./Router.svelte";
export { default as Route } from "./Route.svelte";
export { default as Link } from "./Link.svelte";
export { createLink, createLinks } from "./actions";
export {
	useLocation,
	useResolve,
	useResolvable,
	useNavigate,
	useMatch,
	useParams,
	useFocus,
} from "./hooks";
