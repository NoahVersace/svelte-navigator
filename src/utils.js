/*
 * Adapted from https://github.com/reach/router/blob/b60e6dd781d5d3a4bdaaf4de665649c0f6a7e78d/src/lib/utils.js
 *
 * https://github.com/reach/router/blob/master/LICENSE
 */

const paramRegex = /^:(.+)/;

const SEGMENT_POINTS = 4;
const STATIC_POINTS = 3;
const DYNAMIC_POINTS = 2;
const SPLAT_PENALTY = 1;
const ROOT_POINTS = 1;

/**
 * Check if `string` starts with `search`
 * @param {string} string
 * @param {string} search
 * @return {boolean}
 */
export function startsWith(string, search) {
  return string.substr(0, search.length) === search;
}

/**
 * Check if `segment` is a root segment
 * @param {string} segment
 * @return {boolean}
 */
export function isRootSegment(segment) {
  return segment === "";
}

/**
 * Check if `segment` is a dynamic segment
 * @param {string} segment
 * @return {boolean}
 */
export function isDynamic(segment) {
  return paramRegex.test(segment);
}

/**
 * Check if `segment` is a splat
 * @param {string} segment
 * @return {boolean}
 */
export function isSplat(segment) {
  return segment[0] === "*";
}

/**
 * Strip `str` of potential start and end `/`
 * @param {string} str
 * @return {string}
 */
export function stripSlashes(str) {
  return str.replace(/(^\/+|\/+$)/g, "");
}

/**
 * Split up the URI into segments delimited by `/`
 * @param {string} uri
 * @return {string[]}
 */
export function segmentize(uri) {
  return stripSlashes(uri).split("/");
}

/**
 * Score a route depending on how its individual segments look
 * @param {object} route
 * @param {number} index
 * @return {object}
 */
export function rankRoute(route, index) {
  const score = route.default
    ? 0
    : segmentize(route.fullPath).reduce((acc, segment) => {
        let nextScore = acc;
        nextScore += SEGMENT_POINTS;

        if (isRootSegment(segment)) {
          nextScore += ROOT_POINTS;
        } else if (isDynamic(segment)) {
          nextScore += DYNAMIC_POINTS;
        } else if (isSplat(segment)) {
          nextScore -= SEGMENT_POINTS + SPLAT_PENALTY;
        } else {
          nextScore += STATIC_POINTS;
        }

        return nextScore;
      }, 0);

  return { route, score, index };
}

/**
 * Give a score to all routes and sort them on that
 * @param {object[]} routes
 * @return {object[]}
 */
export function rankRoutes(routes) {
  return (
    routes
      .map(rankRoute)
      // If two routes have the exact same score, we go by index instead
      .sort((a, b) => {
        if (a.score < b.score) {
          return 1;
        }
        if (a.score > b.score) {
          return -1;
        }
        return a.index - b.index;
      })
  );
}

/**
 * Ranks and picks the best route to match. Each segment gets the highest
 * amount of points, then the type of segment gets an additional amount of
 * points where
 *
 *  static > dynamic > splat > root
 *
 * This way we don't have to worry about the order of our routes, let the
 * computers do it.
 *
 * A route looks like this
 *
 *  { fullPath, default, value }
 *
 * And a returned match looks like:
 *
 *  { route, params, uri }
 *
 * @param {object[]} routes
 * @param {string} uri
 * @return {?object}
 */
export function pick(routes, uri) {
  let bestMatch;
  let defaultMatch;

  const [uriPathname] = uri.split("?");
  const uriSegments = segmentize(uriPathname);
  const isRootUri = uriSegments[0] === "";
  const ranked = rankRoutes(routes);

  for (let i = 0, l = ranked.length; i < l; i++) {
    const { route } = ranked[i];
    let missed = false;

    if (route.default) {
      defaultMatch = {
        route,
        params: {},
        uri,
      };
      continue;
    }

    const routeSegments = segmentize(route.fullPath);
    const params = {};
    const max = Math.max(uriSegments.length, routeSegments.length);
    let index = 0;

    for (; index < max; index++) {
      const routeSegment = routeSegments[index];
      const uriSegment = uriSegments[index];

      if (typeof routeSegment !== "undefined" && isSplat(routeSegment)) {
        // Hit a splat, just grab the rest, and return a match
        // uri:   /files/documents/work
        // route: /files/* or /files/*splatname
        const splatName = routeSegment === "*" ? "*" : routeSegment.slice(1);

        params[splatName] = uriSegments
          .slice(index)
          .map(decodeURIComponent)
          .join("/");
        break;
      }

      if (typeof uriSegment === "undefined") {
        // URI is shorter than the route, no match
        // uri:   /users
        // route: /users/:userId
        missed = true;
        break;
      }

      const dynamicMatch = paramRegex.exec(routeSegment);

      if (dynamicMatch && !isRootUri) {
        const value = decodeURIComponent(uriSegment);
        params[dynamicMatch[1]] = value;
      } else if (routeSegment !== uriSegment) {
        // Current segments don't match, not dynamic, not splat, so no match
        // uri:   /users/123/settings
        // route: /users/:id/profile
        missed = true;
        break;
      }
    }

    if (!missed) {
      bestMatch = {
        route,
        params,
        uri: `/${uriSegments.slice(0, index).join("/")}`,
      };
      break;
    }
  }

  return bestMatch || defaultMatch || null;
}

/**
 * Check if the `route.fullPath` matches the `uri`.
 * @param {Object} route
 * @param {string} uri
 * @return {?object}
 */
export function match(route, uri) {
  return pick([route], uri);
}

/**
 * Add the query to the pathname if a query is given
 * @param {string} pathname
 * @param {string} [query]
 * @return {string}
 */
export function addQuery(pathname, query) {
  return pathname + (query ? `?${query}` : "");
}

/**
 * Resolve URIs as though every path is a directory, no files. Relative URIs
 * in the browser can feel awkward because not only can you be "in a directory",
 * you can be "at a file", too. For example:
 *
 *  browserSpecResolve('foo', '/bar/') => /bar/foo
 *  browserSpecResolve('foo', '/bar') => /foo
 *
 * But on the command line of a file system, it's not as complicated. You can't
 * `cd` from a file, only directories. This way, links have to know less about
 * their current path. To go deeper you can do this:
 *
 *  <Link to="deeper"/>
 *  // instead of
 *  <Link to=`{${props.uri}/deeper}`/>
 *
 * Just like `cd`, if you want to go deeper from the command line, you do this:
 *
 *  cd deeper
 *  # not
 *  cd $(pwd)/deeper
 *
 * By treating every path as a directory, linking to relative paths should
 * require less contextual information and (fingers crossed) be more intuitive.
 * @param {string} to
 * @param {string} base
 * @return {string}
 */
export function resolve(to, base) {
  // /foo/bar, /baz/qux => /foo/bar
  if (startsWith(to, "/")) {
    return to;
  }

  const [toPathname, toQuery] = to.split("?");
  const [basePathname] = base.split("?");
  const toSegments = segmentize(toPathname);
  const baseSegments = segmentize(basePathname);

  // ?a=b, /users?b=c => /users?a=b
  if (toSegments[0] === "") {
    return addQuery(basePathname, toQuery);
  }

  // profile, /users/789 => /users/789/profile
  if (!startsWith(toSegments[0], ".")) {
    const pathname = baseSegments.concat(toSegments).join("/");
    return addQuery((basePathname === "/" ? "" : "/") + pathname, toQuery);
  }

  // ./       , /users/123 => /users/123
  // ../      , /users/123 => /users
  // ../..    , /users/123 => /
  // ../../one, /a/b/c/d   => /a/b/one
  // .././one , /a/b/c/d   => /a/b/c/one
  const allSegments = baseSegments.concat(toSegments);
  const segments = [];

  allSegments.forEach(segment => {
    if (segment === "..") {
      segments.pop();
    } else if (segment !== ".") {
      segments.push(segment);
    }
  });

  return addQuery(`/${segments.join("/")}`, toQuery);
}

/**
 * Combines the `basepath` and the `path` into one path.
 * @param {string} basepath
 * @param {string} path
 */
export function combinePaths(basepath, path) {
  const barePath =
    path === "/" ? basepath : `${stripSlashes(basepath)}/${stripSlashes(path)}`;
  return `${stripSlashes(barePath)}/`;
}

/**
 * Normalizes a basepath
 *
 * @param {string} path
 * @returns {string}
 *
 * @example
 * normalizePath("base/path/") // -> "/base/path"
 */
export function normalizePath(path) {
  return `/${stripSlashes(path)}`;
}

/**
 * Joins and normalizes multiple path fragments
 *
 * @param  {...string} pathFragments
 * @returns {string}
 */
export function join(...pathFragments) {
  const joinFragment = fragment =>
    segmentize(fragment)
      .filter(Boolean)
      .join("/");
  const joinedSegments = pathFragments.map(joinFragment).join("/");
  return normalizePath(joinedSegments);
}

/**
 * Normalizes a location for consumption by `Route` children and the `Router`.
 * It removes the apps basepath from the pathnam
 * and sets default values for `search` and `hash` properties.
 *
 * @param {Object} location The current global location supplied by the history component
 * @param {string} basepath The applications basepath (i.e. when serving from a subdirectory)
 *
 * @returns The normalized location
 */
export function normalizeLocation(location, basepath) {
  const { pathname, hash = "", search = "", state, key } = location;
  const baseSegments = segmentize(basepath).filter(Boolean);
  const pathSegments = segmentize(pathname).filter(Boolean);
  while (baseSegments.length) {
    if (baseSegments[0] !== pathSegments[0]) {
      throw new Error(
        `<Router> Invalid state: All locations must begin with the basepath "${basepath}", found "${pathname}"`,
      );
    }
    baseSegments.shift();
    pathSegments.shift();
  }
  return {
    pathname: join(...pathSegments),
    hash,
    search,
    state,
    key,
  };
}

/**
 * Decides whether a given `event` should result in a navigation or not.
 * @param {object} event
 */
export function shouldNavigate(event) {
  return (
    !event.defaultPrevented &&
    event.button === 0 &&
    !(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey)
  );
}

/**
 * Checks if a clicked link triggers a navigation to a route from the same application.
 * Returns `false` if link leads to a different website.
 *
 * @param {Element} anchor The `<a />` element
 * @returns {boolean}
 */
export function hostMatches(anchor) {
  const { host } = window.location;
  return (
    anchor.host === host ||
    // svelte seems to kill anchor.host value in ie11, so fall back to checking href
    startsWith(anchor.href, `https://${host}`) ||
    startsWith(anchor.href, `http://${host}`)
  );
}

/**
 * Resolves a link relative to the router basepath and the uri of the current router.
 *
 * @param {string} path The given path, that will be resolved against a router base
 * @param {string} basepath The basepath of the router (i.e. from calling `useBase`)
 * @param {string} routerBase The current router base (i.e. from calling `useRouterBase`)
 * @returns {string} The resolved path
 *
 * @example
 * resolveLink("relative", "/", "/currentBase") // -> "/currentBase/relative"
 * resolveLink("/absolute", "/", "/currentBase") // -> "/absolute"
 * resolveLink("/absolute", "/base", "/currentBase") // -> "/base/absolute"
 */
export function resolveLink(path, basepath, routerBase, appBase) {
  let resolvedLink;
  if (path === "/") {
    resolvedLink = basepath;
  } else if (startsWith(path, "/")) {
    resolvedLink = resolve(stripSlashes(path), basepath);
  } else {
    resolvedLink = resolve(path, routerBase);
  }
  return join(appBase, resolvedLink);
}

/**
 * Create a locally unique id
 *
 * @returns {number} An id
 */
export const createLocalId = (() => {
  let id = 0;
  return () => id++;
})();

/**
 * Create a globally unique id
 *
 * @returns {string} An id
 */
export function createGlobalId() {
  return Math.random()
    .toString(36)
    .substring(2);
}
