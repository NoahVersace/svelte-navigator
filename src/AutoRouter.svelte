<script>
	import { browserHistory, memoryHistory } from "svelte-navigator-history";
	import Router from "./Router.svelte";
	import { isSSR } from "./utils";

	const UNDEF = undefined;
	export let basepath = UNDEF;
	export let url = UNDEF;
	export let primary = UNDEF;
	export let a11y = UNDEF;

	const canUseDOM = !!(
		!isSSR &&
		window.document &&
		window.document.createElement
	);
	// Use memory history in iframes (for example in Svelte REPL)
	const isEmbeddedPage = !isSSR && window.location.origin === "null";
	const history = canUseDOM && !isEmbeddedPage ? browserHistory : memoryHistory;
</script>

<Router {a11y} {basepath} {url} {primary} {history}>
	<slot />
</Router>
