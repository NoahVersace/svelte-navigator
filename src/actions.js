/*
 * Adapted from https://github.com/EmilTholin/svelte-routing
 *
 * https://github.com/EmilTholin/svelte-routing/blob/master/LICENSE
 */

import { shouldNavigate, addListener, isFunction } from "./utils";
import { createLabel, LINKS_ACTION_ID, LINK_ACTION_ID, warn } from "./warning";

const adjustHref = (node, history) => {
	const nodeHref = node.getAttribute("href");
	if (
		nodeHref &&
		nodeHref !== history.createHref(node.dataset.svnvHref || "")
	) {
		// eslint-disable-next-line no-param-reassign
		node.dataset.svnvHref = nodeHref;
		node.setAttribute("href", history.createHref(nodeHref));
	}
};

const walkChildren = (node, cb) => {
	if (!node) return;
	cb(node);
	[...node.childNodes].forEach(child => {
		walkChildren(child, cb);
	});
};

const isNoRoute = node =>
	isFunction(node.hasAttribute) && node.hasAttribute("noroute");

const isLink = node => !!node && node.tagName === "A";

const findClosestLink = element => {
	while (element && !isLink(element)) {
		// eslint-disable-next-line no-param-reassign
		element = element.parentNode;
	}
	return element;
};

const createActionFactory = (getAnchor, setup, labelId) => defaultHistory => (
	node,
	options,
) => {
	let { history = defaultHistory } = options || {};
	const handleClick = event => {
		const anchor = getAnchor(event);
		if (anchor && anchor.target === "" && shouldNavigate(event)) {
			event.preventDefault();
			const to = anchor.dataset.svnvHref;
			history.navigate(to, { replace: anchor.hasAttribute("replace") });
		}
	};

	const unlisten = addListener(node, "click", handleClick);

	setup(node, history);

	let observer;
	if (process.env.NODE_ENV !== "production") {
		observer = new MutationObserver(mutations => {
			mutations.forEach(({ type, target }) => {
				if (type === "attributes" && isLink(target) && !isNoRoute(target)) {
					warn(
						labelId,
						"You cannot update the href attributes when using the " +
							`"${createLabel(labelId)}" action. Consider using a <Link /> ` +
							"component instead.",
						{
							href: target.dataset.svnvHref,
							replace: target.hasAttribute("replace"),
						},
					);
				}
			});
		});
		observer.observe(node, {
			attributes: true,
			attributeFilter: ["href"],
			subtree: true,
		});
	}

	return {
		// eslint-disable-next-line no-shadow
		update(options) {
			history = (options && options.history) || defaultHistory;
			setup(node, history);
		},
		destroy() {
			unlisten();
			if (process.env.NODE_ENV !== "production") {
				observer.disconnect();
			}
		},
	};
};

export const createLink = createActionFactory(
	event => event.currentTarget,
	(node, history) => {
		if (process.env.NODE_ENV !== "production") {
			if (!isLink(node)) {
				const truncate = (str, length) =>
					str.substr(0, length) + (str.length >= length ? " ..." : "");
				warn(
					LINK_ACTION_ID,
					'You can only use the "use:link" action on <a> elements.\n' +
						`Got ${truncate(node.outerHTML, 80)}`,
				);
			}
		}
		adjustHref(node, history);
	},
	LINK_ACTION_ID,
);

export const createLinks = createActionFactory(
	event => {
		const anchor = findClosestLink(event.target);
		if (anchor && !isNoRoute(anchor)) {
			return anchor;
		}
		return null;
	},
	(node, history) => {
		walkChildren(node, child => {
			if (isLink(child) && !isNoRoute(child)) {
				adjustHref(child, history);
			}
		});
	},
	LINKS_ACTION_ID,
);
