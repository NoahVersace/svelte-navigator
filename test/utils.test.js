import { pick, match, resolve, combinePaths, resolveLink } from "../src/utils";
import { testRoutes } from "./testRoutes";

describe("pick", () => {
  it("picks root over dynamic", () => {
    let routes = [
      { meta: "Root", fullPath: "/" },
      { meta: "Dynamic", fullPath: ":foo" },
    ];
    expect(pick(routes, "/").route.meta).toBe("Root");
  });

  it("a bunch of scenarios", () => {
    expect(pick(testRoutes, "/").route.meta).toBe("Root");
    expect(pick(testRoutes, "/groups/main/users/me").route.meta).toBe(
      "MainGroupMe",
    );
    expect(pick(testRoutes, "/groups/123/users/456").route.meta).toBe(
      "GroupUser",
    );
    expect(pick(testRoutes, "/one/two/three/four/five").route.meta).toBe(
      "Fiver",
    );
    expect(pick(testRoutes, "/groups/main/users").route.meta).toBe(
      "MainGroupUsers",
    );
    expect(pick(testRoutes, "/groups/123/users").route.meta).toBe("GroupUsers");
    expect(pick(testRoutes, "/groups/main").route.meta).toBe("MainGroup");
    expect(pick(testRoutes, "/groups/123").route.meta).toBe("Group");
    expect(pick(testRoutes, "/groups").route.meta).toBe("Groups");
    expect(pick(testRoutes, "/files/some/long/path").route.meta).toBe(
      "FilesDeep",
    );
    expect(pick(testRoutes, "/files").route.meta).toBe("Files");
    expect(pick(testRoutes, "/no/where").route.meta).toBe("Default");
  });

  it("pick /*", () => {
    expect(pick([{ fullPath: "/*" }], "/whatever/else")).toMatchSnapshot();
  });

  it("pick return value", () => {
    expect(pick(testRoutes, "/one/two/three/four/five")).toMatchSnapshot();
  });

  it("splat return value", () => {
    expect(pick(testRoutes, "/files/some/deep/path")).toMatchSnapshot();
  });

  it("dynamic segments + splat return value", () => {
    const routes = [{ fullPath: "/users/:userId/files/*" }];
    const matchedRoute = pick(routes, "/users/ryan/files/some/deep/path");
    expect(matchedRoute.params).toEqual({
      userId: "ryan",
      "*": "some/deep/path",
    });
  });
});

describe("match", () => {
  it("creates correct route data", () => {
    const matchedRoute = match({ fullPath: "/foo", meta: "Foo" }, "/foo");
    expect(matchedRoute.route.meta).toBe("Foo");
    expect(matchedRoute.params).toEqual({});
    expect(matchedRoute.uri).toBe("/foo");
  });

  it("matches params", () => {
    const matchedRoute = match({ fullPath: "/foo/:bar" }, "/foo/hello");
    expect(matchedRoute.params).toEqual({ bar: "hello" });
    expect(matchedRoute.uri).toBe("/foo/hello");
  });

  it("matches splat", () => {
    const matchedRoute = match({ fullPath: "/foo/*" }, "/foo/hello");
    expect(matchedRoute.params).toEqual({ "*": "hello" });
    expect(matchedRoute.uri).toBe("/foo");
  });

  it("matches named splat", () => {
    const matchedRoute = match({ fullPath: "/foo/*splat" }, "/foo/hello");
    expect(matchedRoute.params).toEqual({ splat: "hello" });
    expect(matchedRoute.uri).toBe("/foo");
  });

  it("strips search", () => {
    const paramRoute = match({ fullPath: "/foo/:bar" }, "/foo/hello?baz=123");
    expect(paramRoute.uri).toBe("/foo/hello");
    const splatRoute = match({ fullPath: "/foo/*" }, "/foo/hello?baz=123");
    expect(splatRoute.uri).toBe("/foo");
    const namedRoute = match({ fullPath: "/foo/*splat" }, "/foo/hello?baz=123");
    expect(namedRoute.uri).toBe("/foo");
  });
});

describe("resolve", () => {
  it("works with a / base", () => {
    expect(resolve("contacts/ryan", "/")).toEqual("/contacts/ryan");
  });

  it("a bunch of stuff", () => {
    expect(resolve("/somewhere/else", "/users/123")).toEqual("/somewhere/else");
    expect(resolve("settings", "/users/123")).toEqual("/users/123/settings");
    expect(resolve("../../one/../two/.././three", "/a/b/c/d/e/f/g")).toEqual(
      "/a/b/c/d/e/three",
    );
    expect(resolve("./", "/users/123")).toEqual("/users/123");
    expect(resolve(".", "/users/123")).toEqual("/users/123");
    expect(resolve("../", "/users/123")).toEqual("/users");
    expect(resolve("../..", "/users/123")).toEqual("/");
    expect(resolve("../.././3", "/u/1/g/4")).toEqual("/u/1/3");
    expect(resolve("../.././s?beef=boof", "/u/1/g/4")).toEqual(
      "/u/1/s?beef=boof",
    );
    expect(resolve("../.././3", "/u/1/g/4?beef=boof")).toEqual("/u/1/3");
    expect(resolve("stinky/barf", "/u/1/g/4")).toEqual("/u/1/g/4/stinky/barf");
    expect(resolve("?some=query", "/users/123?some=thing")).toEqual(
      "/users/123?some=query",
    );
    expect(resolve("/groups?some=query", "/users?some=thing")).toEqual(
      "/groups?some=query",
    );
  });
});

describe("combinePaths", () => {
  it("returns basepath", () => {
    expect(combinePaths("/base", "/")).toBe("base/");
  });

  it("removes leading slash", () => {
    expect(combinePaths("/base", "/path")).toMatch(/^base.*/);
  });

  it("adds trailing slash", () => {
    expect(combinePaths("/base", "/path")).toMatch(/.*\/$/);
  });

  it("combines two paths", () => {
    expect(combinePaths("/base", "/path")).toBe("base/path/");
  });

  it("combines complex paths", () => {
    expect(combinePaths("/base/path", "/next/path/to/somewhere")).toBe(
      "base/path/next/path/to/somewhere/",
    );
  });

  it("removes unneccessary slashes", () => {
    expect(combinePaths("/base/", "/path/")).toBe("base/path/");
  });
});

describe("resolveLink", () => {
  it("relative", () => {
    expect(resolveLink("relative", "/", "/currentBase")).toBe(
      "/currentBase/relative",
    );
  });

  it("relative, no slashes", () => {
    expect(resolveLink("relative", "", "currentBase")).toBe(
      "/currentBase/relative",
    );
  });

  it("absolute", () => {
    expect(resolveLink("/absolute", "/", "/currentBase")).toBe("/absolute");
  });

  it("absolute, no slashes", () => {
    expect(resolveLink("/absolute", "/", "currentBase")).toBe("/absolute");
  });

  it("absolute with base", () => {
    expect(resolveLink("/absolute", "/base", "/currentBase")).toBe(
      "/base/absolute",
    );
  });

  it("absolute with base ..", () => {
    expect(resolveLink("../relative", "/base", "/currentBase")).toBe(
      "/relative",
    );
  });

  it("absolute ..", () => {
    expect(resolveLink("../relative", "/", "/currentBase")).toBe("/relative");
  });
});
