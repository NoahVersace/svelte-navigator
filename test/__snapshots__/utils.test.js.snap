// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pick pick /* 1`] = `
Object {
  "params": Object {
    "*": "whatever/else",
  },
  "route": Object {
    "fullPath": "/*",
  },
  "uri": "/",
}
`;

exports[`pick pick return value 1`] = `
Object {
  "params": Object {
    "five": "five",
    "four": "four",
    "one": "one",
    "three": "three",
    "two": "two",
  },
  "route": Object {
    "fullPath": "/:one/:two/:three/:four/:five",
    "meta": "Fiver",
  },
  "uri": "/one/two/three/four/five",
}
`;

exports[`pick splat return value 1`] = `
Object {
  "params": Object {
    "*": "some/deep/path",
  },
  "route": Object {
    "fullPath": "/files/*",
    "meta": "FilesDeep",
  },
  "uri": "/files",
}
`;
